### PR TITLE
 Add value witnesses for single payload enums 

### PR DESCRIPTION
**Pull request by aschwaighofer**
So far single payload enums were implemented in terms of runtime functions which
internally emitted several calls to value witnesses.

This commit adds value witnesses to get and store the enum tag side stepping the
need for witness calls as this information is statically available in many cases

/// int (*getEnumTagSinglePayload)(const T* enum, UINT_TYPE emptyCases)
/// Given an instance of valid single payload enum with a payload of this
/// witness table's type (e.g Optional<ThisType>) , get the tag of the enum.

/// void (*storeEnumTagSinglePayload)(T* enum, INT_TYPE whichCase,
///                                   UINT_TYPE emptyCases)
/// Given uninitialized memory for an instance of a single payload enum with a
/// payload of this witness table's type (e.g Optional<ThisType>), store the
/// tag.

A simple 'for element in array' loop in generic code operating on a
ContiguousArray of Int is ~25% faster on arm64.

rdar://31408033

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test linux

**Comment by jrose-apple** [comment]
What's the code size impact for this change?

**Comment by aschwaighofer** [comment]
I don’t have numbers yet. it should somewhat even out with the removal of the array witnesses (of which there where four)

**Comment by aschwaighofer** [comment]
@swift-ci Please benchmark

File: lib/IRGen/GenType.cpp [hunk 1]
```diff
@@ -340,6 +341,385 @@ FixedTypeInfo::getSpareBitExtraInhabitantIndex(IRGenFunction &IGF,
...
+  //   // Assume that one tag bit is enough if the precise calculation overflows
+  //   // an int32.
+  //   numTags += 1;
+  // else {
+  //   unsigned bits = size * 8U;
+  //   unsigned casesPerTagBitValue = 1U << bits;
+  //   numTags += ((emptyCases + (casesPerTagBitValue - 1U)) >> bits);
+  // }
+  // return (numTags < 256 ? 1 :
+	// 				 numTags < 65536 ? 2 : 4);
```
Comment (slavapestov) [inline lib/IRGen/GenType.cpp]: Indentation

File: lib/IRGen/GenType.cpp [hunk 2]
```diff
@@ -340,6 +341,385 @@ FixedTypeInfo::getSpareBitExtraInhabitantIndex(IRGenFunction &IGF,
...
+  auto *isLT65536 =
+      Builder.CreateICmpULT(numTags, llvm::ConstantInt::get(int32Ty, 65536U));
+  numTags = Builder.CreateSelect(isLT65536, two, four);
+  Builder.CreateBr(returnBB);
+
+  Builder.emitBlock(returnBB);
+  auto *phi = Builder.CreatePHI(int32Ty, 3);
+  phi->addIncoming(one, entryBB);
+  phi->addIncoming(numTags, notLT256BB);
+	return phi;
```
Comment (slavapestov) [inline lib/IRGen/GenType.cpp]: Indentation

**Review by slavapestov — COMMENTED** [review]
It looks like the extra inhabitant value witnesses are now only used when you have an enum whose payload is a structural type, like a tuple. Can you think of a clever way around this, implementing the single payload value witnesses for tuples in terms of the single payload value witnesses of the payload type? If so, you could trim the value witness table further.

**Comment by slavapestov** [comment]
For example, would it be sufficient to say that

- enum tag of a tuple => get enum tag of first tuple component
- store tag into a tuple => store tag into first component

This relies on the fact that extra inhabitants of a tuple type are always the extra inhabitants of its first component. If this were to change, the above would no longer work, but maybe we can live with that restriction?

**Comment by aschwaighofer** [comment]
@slavapestov They are also used for non-fixed size structs and enums. Where the enums are the hairy part. structs currently follow the same rule as tuples (first field's extra inhabitants).

- Singleton enums have the extra inhabitant count of the singleton payload or 0 if the singleton is empty.
This can follow the scheme for structs/tuples:

```
if (hasPayload)
  PayloadT->getEnumTagSinglePayload
```

- NoPayload enums have the number of all values above their greatest discriminator value that fit inside their storage size.
Not dependent on a T. Continues to work.

- SinglePayloadEnums reuse their payload's extra inhabitants minus the number of tags they need for their own discriminator.
I believe that we would have to give up on reusing the payload's extra inhabitants.

- MultiPayloadEnums don't currently provide extra inhabitants

If we are okay to loose the single payload enums extra inhabitants I believe the scheme would work.

A further complication is that now the enum witnesses would have to be described in terms of an abstract size such that we can find the address of the extra tag bytes when we forward to the payload of an enum/first element of a struct/tuple . This is doable (see below) but means that the witnesses of fixed size types will no longer simplify as well because the size is dynamic.

The witnesses would get an additional argument that is the size of the 'outermost' type that we append any potential .
```
getEnumTagSinglePayload(const T *enum, UINT_TYPE emptyCases, const Metatype * self)
```
would become

```
getEnumTagSinglePayload(const T *enum, UINT_TYPE emptyCases, const Metatype * self,const Metatype * outermostType)
```

We pass null at the top-level of witness calls. So fixed size types will still use the static size in the common case.

Hand-wavy sketch for a fixed size type:

```
int getEnumTagSinglePayload(const T *enum, UINT_TYPE emptyCases, const Metatype * self,const Metatype * outermostType) {
   auto size = #FixedSize
   if (outermostType)
      size = outermostType->size()
}
```

The witness for a tuple will look like:

```
int getEnumTagSinglePayload(const T *enum, UINT_TYPE emptyCases, const Metatype * self,const Metatype * typeForSize) {
   auto &metatype = *(const TupleTypeMetadata*) self;
   auto &firstEltTy = metatype.getElement(0).Type;
   firstEltTy->vw_getEnumTagSinglePayload(enum, emptyCases, firstEltTy, self)
}
```

What does that mean for code size: We have the cost of passing an extra zero at all call-sites plus bigger witness implementations for fixed size types because the size is dynamic. Not immediately clear that this is a code size win. You only pay for extra inhabitant witnesses if there are extra inhabitants.

**Comment by slavapestov** [comment]
Ok, we can stick with the current approach then.

**Comment by aschwaighofer** [comment]
The standard library text size increases by 0.9% on x86-64. From 4254426 to 4292298 bytes.

Out of the 37K bytes, 21K bytes are from the witnesses, the rest is overhead because we now have to load the witness on the caller side.

**Comment by aschwaighofer** [comment]
(Numbers on arm are similar)

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test mac os x

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test linux

**Comment by aschwaighofer** [comment]
@swift-ci Please benchmark

**Comment by aschwaighofer** [comment]
Code size swift-source-compat-suite. On interesting fact I noticed while looking at:

```
CleanroomLogger/watchos                    :     187695     193947   -3.33
```

that the delta (3.3%) could roughly be half that. There are duplicate value witnesses that llvm's merge function pass would merge. We run -wmo -num-threads 1..N and so these witnesses are split up per source file and don't get merged. 

That is -wmo -num-threads >0 is not helpful for the merge function pass and we leave code size improvements on the table (not only for witnesses). This is the obvious trade-off of num-threads but had not occurred to me before.

```
Project/platform                           :        Old        New   Percent                                                                                                                                                               
                                                                                                                                                                                                                                           
Alamofire/iphoneos/arm64                   :     539932     545940   -1.11
Alamofire/iphoneos/armv7                   :     562731     568043   -0.94
iOS Example/iOS/arm64                      :      68452      68804   -0.51
iOS Example/iOS/armv7                      :      68454      68766   -0.46
Alamofire/iphoneos                         :    1102663    1113983   -1.03
Alamofire/iphoneos                         :    1102663    1113983   -1.03
iOS Example/iphoneos                       :     136906     137570   -0.49
AMScrollingNavbar/iphoneos/arm64           :      63452      63916   -0.73
AMScrollingNavbar/iphoneos/armv7           :      63054      63478   -0.67
AMScrollingNavbar/iphoneos                 :     126506     127394   -0.70
Socket/iphoneos/arm64                      :     105681     107893   -2.09
Socket/iphoneos/armv7                      :     120081     122081   -1.67
Socket/iphoneos                            :     225762     229974   -1.87
ChattoApp/iphoneos/arm64                   :     168833     169377   -0.32
ChattoApp/iphoneos/armv7                   :     175202     175738   -0.31           
Chatto/iphoneos/arm64                      :     274442     278230   -1.38
Chatto/iphoneos/armv7                      :     285867     289331   -1.21
ChattoAdditions/iphoneos/arm64             :     627469     635153   -1.22
ChattoAdditions/iphoneos/armv7             :     644250     651410   -1.11
Chatto/iphoneos                            :     560309     567561   -1.29
ChattoAdditions/iphoneos                   :    1271719    1286563   -1.17
ChattoApp/iphoneos                         :     344035     345115   -0.31
Chatto/iphoneos                            :     560309     567561   -1.29
ChattoAdditions/iphoneos                   :    1271719    1286563   -1.17
CleanroomLogger/watchos                    :     187695     193947   -3.33
Dollar/macos                               :     186168     188580   -1.30
Dwifft/iphoneos/arm64                      :     105665     106753   -1.03
Dwifft/iphoneos/armv7                      :     113710     114598   -0.78
Dwifft/iphoneos                            :     219375     221351   -0.90
IBAnimatable/iphoneos/arm64                :    1808681    1818465   -0.54
IBAnimatable/iphoneos/armv7                :    1856317    1865133   -0.47
IBAnimatableApp/iphoneos/arm64             :     210223     210715   -0.23
IBAnimatableApp/iphoneos/armv7             :     214435     214911   -0.22
IBAnimatable/iphoneos                      :    3664998    3683598   -0.51
IBAnimatable/iphoneos                      :    3664998    3683598   -0.51
IBAnimatableApp/iphoneos                   :     424658     425626   -0.23
JSQCoreDataKit/iphoneos/arm64              :      79874      81514   -2.05
JSQCoreDataKit/iphoneos/armv7              :      83412      84904   -1.79
JSQCoreDataKit/iphoneos                    :     163286     166418   -1.92
JSQDataSourcesKit/iphoneos/arm64           :     105441     106721   -1.21
JSQDataSourcesKit/iphoneos/armv7           :     102864     103856   -0.96
JSQDataSourcesKit/iphoneos                 :     208305     210577   -1.09
KeychainAccess/watchos                     :     201444     203676   -1.11
Prelude/appletvos                          :     112918     115514   -2.30
Prelude_UIKit/appletvos                    :     252654     256090   -1.36
ReactiveExtensions/appletvos               :     179366     179842   -0.27
ReactiveSwift/appletvos                    :     813594     820886   -0.90
Result/appletvos                           :      61346      61954   -0.99
Kingfisher/watchos                         :     266330     271030   -1.76
Kommander/watchos                          :      67136      67872   -1.10
LaunchScreenSnapshot/iphoneos/arm64        :      44892      45356   -1.03
LaunchScreenSnapshot/iphoneos/armv7        :      45056      45480   -0.94
LaunchScreenSnapshot/iphoneos              :      89948      90836   -0.99
ObjectMapper/watchos                       :     240815     242167   -0.56
Example/iphoneos/arm64                     :      29830      30022   -0.64
Example/iphoneos/armv7                     :      27051      27219   -0.62
PanelKit/iphoneos/arm64                    :     141982     143162   -0.83
PanelKit/iphoneos/armv7                    :     149354     150434   -0.72
Example/iphoneos                           :      56881      57241   -0.63
PanelKit/iphoneos                          :     291336     293596   -0.78
PanelKit/iphoneos                          :     291336     293596   -0.78
PromiseKit/appletvos                       :     263636     264628   -0.38
ReLax/appletvos                            :     156584     159652   -1.96
ReactiveCocoa/watchos                      :      91640      91960   -0.35
ReactiveSwift/watchos                      :     711774     719314   -1.06
Result/watchos                             :      49389      49945   -1.13
ReactiveSwift/watchos                      :     711662     719202   -1.06
Result/watchos                             :      49389      49945   -1.13
Realm/watchos                              :    3118605    3118605    0.00
ReSwift/watchos                            :      47853      48641   -1.65  
RxCocoa/iphoneos/arm64                     :     525336     529820   -0.85
RxCocoa/iphoneos/armv7                     :     484947     489203   -0.88
RxSwift/iphoneos/arm64                     :     900068     909924   -1.10
RxSwift/iphoneos/armv7                     :     850946     861450   -1.23
RxCocoa/iphoneos                           :    1010283    1019023   -0.87
RxSwift/iphoneos                           :    1751014    1771374   -1.16
RxSwift/appletvos                          :    1087784    1099264   -1.06
RxTest/appletvos                           :      78743      80039   -1.65
Siesta/macos                               :     603334     611178   -1.30
SiestaUI/macos                             :         48         48    0.00
Starscream/appletvos                       :     129709     131025   -1.01
SwiftDate/watchos                          :     318961     321833   -0.90
SwifterSwift/watchos                       :     414704     416216   -0.36
                                                                          
Total                                      :   40590274   40923954   -0.82
```

**Pull request merged**